google-cloud-firestore/test/google/cloud/firestore/bulkwriter_test.rb

```diff
@@ -0,0 +1,333 @@
+      requests = [request_1, request_2]
+      stub = BatchWriteStub.new responses, requests
+      firestore.service.instance_variable_set :@firestore, stub
+
+      bw = firestore.bulk_writer
+
+      result_1 = bw.create "cities/NYC", { foo: "bar"}
+      result_2 = bw.create "cities/MTV", { foo: "bar"}
+      bw.flush
+      bw.close
+
+      _(result_1.value).must_be_kind_of Google::Cloud::Firestore::BulkWriterOperation::WriteResult
+      _(result_2.value).must_be_kind_of Google::Cloud::Firestore::BulkWriterOperation::WriteResult
+      _(stub.responses.length).must_equal 0
+      _(stub.requests.length).must_equal 0
+    end
+
+    it "retry operations in correct order" do
+      write_1 = Google::Cloud::Firestore::Convert.write_for_create("#{documents_path}/cities/NYC", { foo: "bar"})
+      write_2 = Google::Cloud::Firestore::Convert.write_for_create("#{documents_path}/cities/MTV", { foo: "bar"})
+      request_1 = batch_write_args [write_1]
+      request_2 = batch_write_args [write_2]
+      responses = [batch_write_fail_resp(1), batch_write_fail_resp(1), batch_write_fail_resp(1), batch_write_pass_resp(1), batch_write_pass_resp(1)]
+      requests = [request_1, request_1, request_2, request_2, request_1]
+      stub = BatchWriteStub.new responses, requests
+      firestore.service.instance_variable_set :@firestore, stub
+
+      bw = firestore.bulk_writer
+
+      result_1 = bw.create "cities/NYC", { foo: "bar"}
+      sleep 1.5
+      result_2 = bw.create "cities/MTV", { foo: "bar"}
+      bw.flush
+      bw.close
+
+      _(result_1.value).must_be_kind_of Google::Cloud::Firestore::BulkWriterOperation::WriteResult
+      _(result_2.value).must_be_kind_of Google::Cloud::Firestore::BulkWriterOperation::WriteResult
+      _(stub.responses.length).must_equal 0
+      _(stub.requests.length).must_equal 0
+    end
+  end
+
+  describe "thread tests" do
+
+    it "request threads tests" do
+      write_1 = Google::Cloud::Firestore::Convert.write_for_create("#{documents_path}/cities/NYC", { foo: "bar"})
+      write_2 = Google::Cloud::Firestore::Convert.write_for_create("#{documents_path}/cities/MTV", { foo: "bar"})
+      write_3 = Google::Cloud::Firestore::Convert.write_for_create("#{documents_path}/cities/KIR", { foo: "bar"})
+      requests = [batch_write_args([write_1, write_2, write_3])]
+      responses = [batch_write_pass_resp(3)]
+      stub = BatchWriteStub.new responses, requests, 1
+      firestore.service.instance_variable_set :@firestore, stub
+
+      thread_count = Thread.list.count
+      bw = firestore.bulk_writer request_threads: 1, batch_threads: 2
+
+      result_1 = bw.create "cities/NYC", { foo: "bar"}
+      result_2 = bw.create "cities/MTV", { foo: "bar"}
+      result_3 = bw.create "cities/KIR", { foo: "bar"}
+      sleep 0.1
+      thread_count_2 = Thread.list.count
+      bw.flush
+      bw.close
+
+      _(thread_count_2 - thread_count).must_be :<=, 3
```

Comment:
Why <= and not =?

Reply:
It is a mistake. `:==` should be used for a tighter bound instead of `:<=`.